### DISCUSSION TITLE
SourceGenerator: Is there a way to include files from a dependant project?

### DISCUSSION THREAD

**Discussion by dotlogix** [dotnet/roslyn#64692] [Q&A]
Hey there I am currently testing SourceGenerators.

My main goal is to create DTOs from entities using source generation.
My problem is now that it seems like I can't analyze files from another project.

I have two projects one for Adapters (Dtos, ApiClients, etc) and another one for the Server (Controllers, Services, Entities).
The server references the Adapter project to access the Dtos in the end.

So because of the dependencies I think I can not get the SyntaxNodes of my Entities from a SourceGenerator added to my Adapter project.
I also tried linking the Entity.cs files, but roslyn does not pick these up as SourceFiles for Analysis because they are linked as AdditionalFiles.

Also because my generator creates DTOs and my Server uses these DTOs I can not do dll linking.
Seperating the entities in it's own Dll would also not solve the problem because then my Adapter project would need a reference to the Entity project.

So is there any way you can think of to access the Entity SyntaxTree information from my DTO project?

**Comment by taori** [comment]
Unless you try adding a reference to the other projects files as existing files link you are probably not going to get the syntax information. Keep in mind that what you are talking about sounds like what you want to be doing should be done by a source generator, but may ruin the performance if you cannot control when it runs (i.e. with every code change)

Have you considered writing a CLI application utilizing roslyn as a driver for your generation demands? Effectively you could spawn your own workspace using the solution. Then you have full context of all syntax trees and do whatever you like. It's not the built in experience with updates on source change though and you lose all benefits from caching.

**Reply by dotlogix** [reply]
Thx for your answer, yeah propably this would be the safest approach, but doing that would be far out of scope of my current tasks.

I settled down with sth similar to this and use additional files as linked shared cs files. You don't get all of the type information but if you only need it for some simple classes like Entities and Relations this works out :)

``` csharp
[Generator]
public class DtoGenerator : IIncrementalGenerator {
	public void Initialize(IncrementalGeneratorInitializationContext context) {
		var entitySyntaxTrees = context.AdditionalTextsProvider
		   .Where(text => text.Path.EndsWith("Entity.cs"))
		   .Select((text, cancellationToken) => CSharpSyntaxTree.ParseText(text.GetText(cancellationToken)));

		var entitiesToConsider = entitySyntaxTrees
		   .Combine(entitySyntaxTrees.Collect())
		   .Combine(context.CompilationProvider)
		   .SelectMany((tuple, cancellationToken) => {
				var ((syntaxTree, syntaxTrees), compilation) = tuple;
				compilation = compilation.AddSyntaxTrees(syntaxTrees);
				return syntaxTree.GetRoot()
				   .DescendantNodes()
				   .Where(node => FilterNode(node, cancellationToken))
				   .Select(node => (node, semanticModel: compilation.GetSemanticModel(syntaxTree), compilation));
			})
		   .Select((tuple, token) => (entityType: TransformNode(tuple.node, tuple.semanticModel, token), tuple.compilation));

		context.RegisterSourceOutput(entitiesToConsider, GenerateCode);
	}

	private static bool FilterNode(SyntaxNode node, CancellationToken cancellationToken) {
		return node is ClassDeclarationSyntax cds && cds.Identifier.Text.EndsWith("Entity");
	}

	private static EntityTypeInfo TransformNode(SyntaxNode node, SemanticModel semanticModel, CancellationToken cancellationToken) {
		return new EntityTypeInfo(
			(ClassDeclarationSyntax)node,
			(INamedTypeSymbol)semanticModel.GetDeclaredSymbol(node)
		);
	}

	private static void GenerateCode(SourceProductionContext context, (EntityTypeInfo entityType, Compilation compilation) args) {
		var (entityType, _) = args;
		context.AddSource($"{entityType.BaseName}Query_Test.g.cs", SourceText.From(
			$@"public partial class {entityType.BaseName}Response {{}}", Encoding.UTF8
		));
	}

	private class EntityTypeInfo {
		public ClassDeclarationSyntax Node { get; }
		public INamedTypeSymbol Symbol { get; }

		public string Name { get; }
		public string BaseName { get; set; }

		public EntityTypeInfo(ClassDeclarationSyntax node, INamedTypeSymbol symbol) {
			Node = node;
			Symbol = symbol;
			
			Name = Symbol.Name;
			BaseName = Name.Remove(Name.Length - "Entity".Length);
		}
	}
}
```
